### PR TITLE
Change discovery service to lookup scopes index

### PR DESCRIPTION
**Pull request by naveenreddykarri**
Change discovery service to lookup scopes index
@nkunal @dilipdevaraj-sfdc

### CODE REVIEW

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 1]
```diff
@@ -66,6 +66,9 @@
 @Singleton
 public class ElasticSearchSchemaService extends AbstractSchemaService {
 
+	private static final String SCOPE_INDEX_NAME = "scopes";
```
[the remark this answers is not in the document]
Comment (dilipdevaraj-sfdc) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: @naveenreddykarri 
You can leverage https://github.com/salesforce/Argus/pull/812 which is doing the indexing part. Some things that you need for the search part might be common.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java [hunk 2]
```diff
@@ -170,6 +173,20 @@ public MetricSchemaRecord getScanFrom() {
 		return scanFrom;
 	}
 
+	public boolean QueryOnlyOnScope() {
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java]: pls follow Java camel case naming conventions

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 3]
```diff
@@ -344,16 +347,27 @@ protected void implementationSpecificPut(List<Metric> metrics) {
 		Map<String, String> tags = new HashMap<>();
 		tags.put("type", "REGEXP_WITH_AGGREGATION");
 		long start = System.currentTimeMillis();
+
+		String index_name = INDEX_NAME;
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: again pls follow Java naming conventions

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 4]
```diff
@@ -344,16 +347,27 @@ protected void implementationSpecificPut(List<Metric> metrics) {
...
+		String type_name = TYPE_NAME;
+
+		if (query.QueryOnlyOnScope() && RecordType.SCOPE.equals(type))
+		{
+			index_name = SCOPE_INDEX_NAME;
+			type_name = SCOPE_TYPE_NAME;
+		}
+
 		String requestUrl = new StringBuilder().append("/")
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Were you able to successfully read from you scope only index by testing it from metric discovery page by typing in something only in scope field ?

Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: yep, I checked with existing 'scopes' index.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java (hunk 2 above)
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java]: fixed.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java [hunk 5]
```diff
@@ -170,6 +173,20 @@ public MetricSchemaRecord getScanFrom() {
...
+
+		if (!Objects.equals(this.scope, REGEX_MATCH_ANY) &&
```
[the remark this answers is not in the document]
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java]: fixed.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java (hunk 3 above)
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: fixed

**Comment by naveenreddykarri** [comment]
added unit tests

File: ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java [hunk 6]
```diff
@@ -170,6 +173,19 @@ public MetricSchemaRecord getScanFrom() {
 		return scanFrom;
 	}
 
+	public boolean queryOnlyOnScope() {
+
+		if (Objects.equals(this.namespace, REGEX_MATCH_ANY) &&
+				Objects.equals(this.metric, REGEX_MATCH_ANY) &&
+				Objects.equals(this.tagKey, REGEX_MATCH_ANY) &&
+				Objects.equals(this.tagValue, REGEX_MATCH_ANY))
+		{
```
Comment (nkunal) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java]: Could you fix the spacing here.

File: .gitignore [hunk 7]
```diff
@@ -38,6 +38,8 @@ local.properties
 # Intellij
 .idea/
 *.iml
+argus.ipr
```
Comment (dilipdevaraj-sfdc) [inline .gitignore]: If this is gitignore specific for Intellij
I would suggest *.ipr, *.iws like L40

File: ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java [hunk 8]
```diff
@@ -170,6 +173,19 @@ public MetricSchemaRecord getScanFrom() {
 		return scanFrom;
 	}
 
+	public boolean queryOnlyOnScope() {
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/entity/MetricSchemaRecordQuery.java]: pls rename to isQueryOnlyOnScope

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+        String requestUrl = requestUrlCaptor.getValue();
+
+        assertEquals("/scopenames/scope_type/_search", requestUrl);
+
+        assertTrue(queryForScope.queryOnlyOnScope());
+    }
+
+    @Test
+    public void testGetUniqueUsingMetricIndex() throws IOException {
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java]: rename method to testGetUniqueUsingMetricSchemaIndex

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+            "\t\t\"max_score\": 0.0,\n" +
+            "\t\t\"hits\": []\n" +
+            "\t},\n" +
+            "\t\"aggregations\": {\n" +
+            "\t\t\"distinct_values\": {\n" +
+            "\t\t\t\"doc_count_error_upper_bound\": 0,\n" +
+            "\t\t\t\"sum_other_doc_count\": 424,\n" +
+            "\t\t\t\"buckets\": [\n" +
+            "\t\t\t\t{\n" +
+            "\t\t\t\t\t\"key\": \"system.PHX.AGG.insights-phx-sp1\\n\",\n" +
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java]: please remove specific key (system....) and make it generic like localhost

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            "\t\t\t\t\t\"doc_count\": 1\n" +
+            "\t\t\t\t},\n" +
+            "\t\t\t\t{\n" +
+            "\t\t\t\t\t\"key\": \"system.PHX.AGG.insights-phx-sp2\\n\",\n" +
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java]: same here

File: ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+        assertEquals("/scopenames/scope_type/_search", requestUrl);
+
+        assertTrue(queryForScope.queryOnlyOnScope());
+    }
+
+    @Test
+    public void testGetUniqueUsingMetricIndex() throws IOException {
+
+        MetricSchemaRecordQuery queryForScope = new MetricSchemaRecordQuery.MetricSchemaRecordQueryBuilder().scope("system")
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/test/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaServiceTest.java]: queryForScope is wrong name here, when you are querying entire metrics schema

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 13]
```diff
@@ -396,16 +397,27 @@ protected void implementationSpecificPut(List<Metric> metrics, List<String> scop
 		Map<String, String> tags = new HashMap<>();
 		tags.put("type", "REGEXP_WITH_AGGREGATION");
 		long start = System.currentTimeMillis();
+
+		String indexName = INDEX_NAME;
+		String typeName = TYPE_NAME;
+
+		if (query.isQueryOnlyOnScope() && RecordType.SCOPE.equals(type))
+		{
+			indexName = SCOPE_INDEX_NAME;
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: In this approach how are we differentiating the auditing for 
scopenames.query.count
scopenames.query.latency - (time series for scopenames query latency) 
when we are merging the returned results into MetricSchemaRecord ?

@nkunal @naveenreddykarri

Comment (nkunal) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: we will have a separate story to track the count/latency for the new index/s

Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: fixed.

**Pull request merged**